llvm/lib/MC/GOFFObjectWriter.cpp

```diff
@@ -217,34 +222,403 @@ void GOFFOstream::write_impl(const char *Ptr, size_t Size) {
+                            ? (Symbol.isExported() ? GOFF::ESD_BSC_ImportExport
+                                                   : GOFF::ESD_BSC_Library)
+                            : GOFF::ESD_BSC_Section;
+    if (!Symbol.isExternal())
+      SD.BindingScope = GOFF::ESD_BSC_Section;
+
+    GOFFSymbol ED = createWSASymbol(SD.EsdId);
+    GOFFSymbol PR = createPRSymbol(SectionName, ED.EsdId);
+    ED.Alignment =
+        std::max(static_cast<GOFF::ESDAlignment>(Log2(Section.getAlign())),
+                 GOFF::ESD_ALIGN_Quadword);
+
+    PR.Executable = GOFF::ESD_EXE_DATA;
+    PR.NameSpace = GOFF::ESD_NS_Parts;
+    PR.BindingScope = BindingScope;
+    PR.setAlignment(Symbol.getAlignment());
+    PR.SectionLength = Layout.getSectionAddressSize(&Section);
+
+    // As above, avoid zero-length sections for data.
+    if (!PR.SectionLength)
+      PR.SectionLength = 2;
+
+    writeSymbol(SD, Layout);
+    writeSymbol(ED, Layout);
+    writeSymbol(PR, Layout);
+
+    SymbolMap.insert(std::make_pair(&Symbol, PR.EsdId));
+    GOFFSection GoffSec = GOFFSection(PR.EsdId, PR.EsdId, SD.EsdId);
+    SectionMap.insert(std::make_pair(&Section, GoffSec));
+  } else
+    llvm_unreachable("Unhandled section kind for Symbol");
+}
+
+void GOFFObjectWriter::writeSymbolDeclaredInModule(const MCSymbolGOFF &Symbol,
+                                                   MCAssembler &Asm,
+                                                   const MCAsmLayout &Layout) {
+  GOFFSymbol SD = RootSD;
+
+  GOFF::ESDExecutable Exec = Symbol.getExecutable();
+
+  GOFFSymbol GSym;
+  uint32_t GSymEsdId = 0;
+  switch (Exec) {
+  case GOFF::ESD_EXE_CODE:
+  case GOFF::ESD_EXE_Unspecified: {
+    bool IsIndirectSymbol = Symbol.hasAliasName();
+    GOFFSymbol ER = createERSymbol(IsIndirectSymbol ? Symbol.getAliasName()
+                                                    : Symbol.getName(),
+                                   SD.EsdId, &Symbol);
+    ER.BindingScope = GOFF::ESD_BSC_ImportExport;
+    GSym.Indirect = IsIndirectSymbol;
+    GSym = ER;
+    GSymEsdId = ER.EsdId;
+    break;
+  }
+  case GOFF::ESD_EXE_DATA: {
+    GOFFSymbol ED = createWSASymbol(SD.EsdId);
+    ED.setAlignment(Symbol.getAlignment());
+    writeSymbol(ED, Layout);
+    GOFFSymbol PR = createPRSymbol(Symbol.getName(), ED.EsdId);
+    PR.BindingScope = GOFF::ESD_BSC_ImportExport;
+    PR.setAlignment(Symbol.getAlignment());
+    PR.NameSpace = GOFF::ESD_NS_Parts;
+    PR.SectionLength = 0;
+    GSym = PR;
+    GSymEsdId = PR.EsdId;
+    break;
+  }
+  }
+
+  GSym.BindingScope =
+      Symbol.isExported() ? GOFF::ESD_BSC_ImportExport : GOFF::ESD_BSC_Library;
+
+  SymbolMap.insert(std::make_pair(&Symbol, GSymEsdId));
+  writeSymbol(GSym, Layout);
+}
+
+void GOFFObjectWriter::executePostLayoutBinding(MCAssembler &Asm,
+                                                const MCAsmLayout &Layout) {
+  LLVM_DEBUG(dbgs() << "Entering " << __FUNCTION__ << "\n");
+
+  defineRootSD(Asm, Layout);
```

Comment:
Why does this have to be done here?  Can't this go into e.g. writeADAandCodeSectionSymbols?